lib/iiif/service.rb

```diff
@@ -206,26 +211,31 @@ def self.from_ordered_hash(hsh, default_klass=IIIF::OrderedHash)
 
       hsh.keys.each do |key|
         new_key = key.underscore == key ? key : key.underscore
-        if new_key == 'service'
-          new_object[new_key] = IIIF::Service.from_ordered_hash(hsh[key], IIIF::Service)
-        elsif new_key == 'resource'
-          new_object[new_key] = IIIF::Service.from_ordered_hash(hsh[key], IIIF::Presentation::Resource)
-        elsif hsh[key].kind_of?(Hash)
-          new_object[new_key] = IIIF::Service.from_ordered_hash(hsh[key])
-        elsif hsh[key].kind_of?(Array)
-          new_object[new_key] = []
-          hsh[key].each do |member|
-            if new_key == 'service'
-              new_object[new_key] << IIIF::Service.from_ordered_hash(member, IIIF::Service)
-            elsif member.kind_of?(Hash)
-              new_object[new_key] << IIIF::Service.from_ordered_hash(member)
-            else
-              new_object[new_key] << member
-              # Again, no nested arrays, right?
+        begin
+          if new_key == 'service'
+            new_object[new_key] = IIIF::Service.from_ordered_hash(hsh[key], IIIF::Service)
+          elsif new_key == 'resource'
+            new_object[new_key] = IIIF::Service.from_ordered_hash(hsh[key], IIIF::Presentation::Resource)
+          elsif hsh[key].kind_of?(Hash)
+            new_object[new_key] = IIIF::Service.from_ordered_hash(hsh[key])
+          elsif hsh[key].kind_of?(Array)
+            new_object[new_key] = []
+            hsh[key].each do |member|
+              if new_key == 'service'
+                new_object[new_key] << IIIF::Service.from_ordered_hash(member, IIIF::Service)
+              elsif member.kind_of?(Hash)
+                new_object[new_key] << IIIF::Service.from_ordered_hash(member)
+              else
+                new_object[new_key] << member
+                # Again, no nested arrays, right?
+              end
             end
+          else
+            new_object[new_key] = hsh[key]
           end
-        else
-          new_object[new_key] = hsh[key]
+        rescue ExpectedHashError
```

Comment:
I'd prefer we don't do flow control with exceptions. Can we check that `hsh[key]` and `member` are hashes here?